### PR TITLE
162 feature fix rotation of path vizualization

### LINKED ISSUE #162: [Feature]: Fix rotation of path vizualization
### Description

Show path correctly even when car rotates 

### Definition of Done

- path is shown correctly even when car rotates

### PR DESCRIPTION
**Pull request by ghost**
# Description

 MainFramePublisher now works even if car turns (heading changes).

Only heading change is taken into account. Other rotations are ignored. This will not be a problem as long as car doesn't drive on a tilted plane like up a hill or over a curb. But it should do without this functionality for the moment. 

Fixes #162 

## Time invested

@Julian-Graf 8h

## Type of change

Please delete options that are not relevant.

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)

## Does this PR introduce a breaking change?

No

## Most important changes

- PositionPublisher publishes heading correctly
- MainFramePublisher works even if car turns (heading changes)
- small fix to PurePursuitController to declutter logs
- small fix to CoordinateTransformer

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works (might be obsolete with CI later on)
- [x] New and existing unit tests pass locally with my changes (might be obsolete with CI later on)

### CODE REVIEW

**Review by JoKircher — CHANGES_REQUESTED** [review]
Nice PR, however the changes where not really visible as the car wasn't moving at all. It seems the velocity and acc publisher finshed at the beginning:

```
paf22-agent-1            | [velocity_publisher_dummy-9] process has finished cleanly
paf22-agent-1            | log file: /home/carla/.ros/log/d2b1d132-a867-11ed-850f-0242ac140003/velocity_publisher_dummy-9*.log
paf22-agent-1            | [AccDistancePublisherDummy-13] process has finished cleanly
paf22-agent-1            | log file: /home/carla/.ros/log/d2b1d132-a867-11ed-850f-0242ac140003/AccDistancePublisherDummy-13*.log
```
Is this behaviour expected or was it an mistake on my end?

I was also wondering what the rviz config changes are doing? For me there was no visible difference.

Review time: 30 min

File: code/perception/src/coordinate_transformation.py [hunk 1]
```diff
@@ -120,12 +104,14 @@ def ecef_to_enu(x, y, z, lat0, lon0, h0):
     return xE, yN, zUp
 
 
-def quat2heading(quat: [float, float, float, float]):  # todo: docs
-    if all(v == 0 for v in quat):
-        return [0.0, 0.0, 0.0]
-    rot = R.from_quat(quat)
-    return rot.as_euler("xyz", degrees=True)
-#
+def quat2heading(msg):
```
Comment (schwalga) [inline code/perception/src/coordinate_transformation.py:107]:
```suggestion
def quat_to_heading(msg):
```
Just a small suggestion, to be inline with the current naming scheme

File: code/perception/src/PositionPublisher.py [hunk 2]
```diff
@@ -81,10 +74,8 @@ def update_current_pos(self):
         # x, y, z = self.transformer.gnss_to_xyz(lat, lon, alt)
 
         orientation_quat = self.current_pos_gps.pose.pose.orientation
-        self.current_heading = quat2heading([orientation_quat.x,
-                                            orientation_quat.y,
-                                            orientation_quat.z,
-                                            orientation_quat.w])[0]
+        self.current_heading = quat2heading(orientation_quat)
```
Comment (schwalga) [inline code/perception/src/PositionPublisher.py:77]:
```suggestion
        self.current_heading = quat_to_heading(orientation_quat)
```
Suggestion to be inline with current naming scheme.

**Review by schwalga — CHANGES_REQUESTED** [review]
Great PR that includes a lot of improvements.
Unlike @JoKircher the agent ran without issue, however it looks like the displayed height of the path increases as the vehicle gets going? Is this an issue with the current_pos topic?
Am I the only one that observes this behavior?

30 min @schwalga

File: code/perception/src/PositionPublisher.py (hunk 2 above)
Comment (schwalga) [reply code/perception/src/PositionPublisher.py:77]: The import would also need to be updated.

**Comment by schwalga** [comment]
> Nice PR, however the changes where not really visible as the car wasn't moving at all. It seems the velocity and acc publisher finshed at the beginning:
> 
> ```
> paf22-agent-1            | [velocity_publisher_dummy-9] process has finished cleanly
> paf22-agent-1            | log file: /home/carla/.ros/log/d2b1d132-a867-11ed-850f-0242ac140003/velocity_publisher_dummy-9*.log
> paf22-agent-1            | [AccDistancePublisherDummy-13] process has finished cleanly
> paf22-agent-1            | log file: /home/carla/.ros/log/d2b1d132-a867-11ed-850f-0242ac140003/AccDistancePublisherDummy-13*.log
> ```
> 
> Is this behaviour expected or was it an mistake on my end?
> 
> I was also wondering what the rviz config changes are doing? For me there was no visible difference.
> 
> Review time: 30 min

At the moment the vehicle doesn't receive a target speed.
You would have to change line 24  in `acting.launch` from `value="False"` -> `value="True"`

**Comment by JoKircher** [comment]
> > Nice PR, however the changes where not really visible as the car wasn't moving at all. It seems the velocity and acc publisher finshed at the beginning:
> > ```
> > paf22-agent-1            | [velocity_publisher_dummy-9] process has finished cleanly
> > paf22-agent-1            | log file: /home/carla/.ros/log/d2b1d132-a867-11ed-850f-0242ac140003/velocity_publisher_dummy-9*.log
> > paf22-agent-1            | [AccDistancePublisherDummy-13] process has finished cleanly
> > paf22-agent-1            | log file: /home/carla/.ros/log/d2b1d132-a867-11ed-850f-0242ac140003/AccDistancePublisherDummy-13*.log
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Is this behaviour expected or was it an mistake on my end?
> > I was also wondering what the rviz config changes are doing? For me there was no visible difference.
> > Review time: 30 min
> 
> @JoKircher
> 
> At the moment the vehicle doesn't receive a target speed. You would have to change line 24 in `acting.launch` from `value="False"` -> `value="True"`

Ah, thanks, its now working, I can also observe the path "flying" in the sky.

File: code/acting/src/acting/MainFramePublisher.py [hunk 3]
```diff
@@ -39,27 +47,20 @@ def loop(timer_event=None):
...
-                cos(radians(rot_deg)) * self.current_pos.pose.position.y
-            pos[2] = self.current_pos.pose.position.z
-            pos[0] = pos[0] * -1
-            pos[1] = pos[1] * -1
-            rot_quat = R.from_euler("xyz", [0, 0, rot_deg], degrees=True) \
-                .as_quat()
+                cos(rot) * self.current_pos.pose.position.y
+            pos[2] = -self.current_pos.pose.position.z
+            pos[0] = pos[0] * 1
+            pos[1] = pos[1] * 1
```
[the remark this answers is not in the document]
Comment (Julian-Graf) [reply code/acting/src/acting/MainFramePublisher.py]: Fixed in 9dfd9398486344685369e5cc9cbd6258f74f01cf

File: code/perception/src/EKFTranslationNode.py [hunk 4]
```diff
@@ -77,27 +74,30 @@ def update_imu_data(self, data: Imu):
         imu_data.header.stamp = data.header.stamp
         imu_data.header.frame_id = "hero"
 
-        imu_data.orientation.x = data.orientation.x
-        imu_data.orientation.y = data.orientation.y
-        imu_data.orientation.z = data.orientation.z
-        imu_data.orientation.w = data.orientation.w
+        rot = [data.orientation.x, data.orientation.y,
+               data.orientation.z, data.orientation.w]
```
[the remark this answers is not in the document]
Comment (Julian-Graf) [reply code/perception/src/EKFTranslationNode.py]: Fixed in 9dfd9398486344685369e5cc9cbd6258f74f01cf

File: code/perception/src/PositionPublisher.py (hunk 2 above)
Comment (Julian-Graf) [reply code/perception/src/PositionPublisher.py]: Fixed in 9dfd9398486344685369e5cc9cbd6258f74f01cf

**Comment by Julian-Graf** [comment]
> > > Nice PR, however the changes where not really visible as the car wasn't moving at all. It seems the velocity and acc publisher finshed at the beginning:
> > > ```
> > > paf22-agent-1            | [velocity_publisher_dummy-9] process has finished cleanly
> > > paf22-agent-1            | log file: /home/carla/.ros/log/d2b1d132-a867-11ed-850f-0242ac140003/velocity_publisher_dummy-9*.log
> > > paf22-agent-1            | [AccDistancePublisherDummy-13] process has finished cleanly
> > > paf22-agent-1            | log file: /home/carla/.ros/log/d2b1d132-a867-11ed-850f-0242ac140003/AccDistancePublisherDummy-13*.log
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Is this behaviour expected or was it an mistake on my end?
> > > I was also wondering what the rviz config changes are doing? For me there was no visible difference.
> > > Review time: 30 min
> > 
> > 
> > @JoKircher
> > At the moment the vehicle doesn't receive a target speed. You would have to change line 24 in `acting.launch` from `value="False"` -> `value="True"`
> 
> Ah, thanks, its now working, I can also observe the path "flying" in the sky.

Could you supply a screenshot, so can get a better understanding of whats the problem

**Comment by Julian-Graf** [comment]
> image

Thanks. It looks like a small offset. I'll look into it

**Comment by Julian-Graf** [comment]
This is caused because the Position publisher publishes an incorrect z component (height) of the car. When the car drives on the curb, the z coordinate changes by about 10 m. This leads to the path being displaced. When driving along the road I don't have this problem. Which z-position is published in your simulation (`rostopic echo /carla/hero/current_pos`)?

A workaround for the moment could be changing the z-offset of the path in Rvis

**Comment by JoKircher** [comment]
image

The z coordinate is linearily decreasing from 39 to 28 independent of hitting curbs

**Comment by Julian-Graf** [comment]
I updated the branch and tried out a few things. It does work better than what we have at the moment, but it still has the bug that the path will start to float which has to be corrected using the offset.
I am not sure, if we should merge this or not

Time invested: 30 min

**Review by JoKircher — APPROVED** [review]
The changes look better and this PR will help to follow the through the leaderboard!

**Review by JoKircher — APPROVED** [review]

**Review by schwalga — APPROVED** [review]
Looks good to me.

**Pull request merged**